cis194/week2/zehua/src/Exercises.hs

```diff
@@ -31,3 +32,50 @@ parseMessage s =
 
 parse :: String -> [LogMessage]
 parse = fmap parseMessage . lines
+
+-- ex2
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) t = t
+insert lm Leaf = Node Leaf lm Leaf
+insert lm@(LogMessage _ ts _ ) (Node l lm1@(LogMessage _ ts1 _) r)
+  | ts < ts1 = Node (insert lm l) lm1 r
+  | otherwise = Node l lm1 (insert lm r)
+
+-- ex3
+build :: [LogMessage] -> MessageTree
+build = foldr insert Leaf
+
+-- ex4
+inOrder :: MessageTree -> [LogMessage]
+inOrder Leaf          = []
+inOrder (Node l lm r) = inOrder l ++ [lm] ++ inOrder r
+
+-- ex5
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong = fmap (fromJust . getMessage) . filter error50 . inOrder . build
+  where
+    error50 (LogMessage (Error s) _ _) = s >= 50
+    error50 _                          = False
+    getMessage (LogMessage _ _ m) = Just m
+    getMessage _                  = Nothing
```

Comment:
don't want to use `Maybe` here because we know from the filtering with `error50` that there would be no `Unknown`. But the compiler would complain. :|

Reply:
Folding is what you want here I think.

Reply:
not sure how to fold it. do you mean `concat . F.toList` or `concatMap`?

anyway, it looks like `catMaybes` works better than `fromJust`. I have changed it to use `catMaybes`.